modules/core/src/test/java/org/apache/ignite/internal/processors/odbc/ClientListenerMetricsTest.java

```diff
@@ -0,0 +1,222 @@
+                .setSslClientCertificateKeyStoreType(DFLT_STORE_TYPE)
+                .setSslClientCertificateKeyStorePassword("123456")
+                .setSslTrustCertificateKeyStorePath(GridTestUtils.keyStorePath("trustone"))
+                .setSslTrustCertificateKeyStoreType(DFLT_STORE_TYPE)
+                .setSslTrustCertificateKeyStorePassword("123456");
+
+            try {
+                Ignition.startClient(cfgSsl);
+
+                fail("Should be rejected.");
+            }
+            catch (ClientException ignored) {}
+
+            checkRejectMetrics(mreg, 1, 0, 1);
+
+            ClientConfiguration cfgAuth = getClientConfiguration()
+                .setUserName("SomeRandomInvalidName")
+                .setUserPassword("42");
+
+            try {
+                Ignition.startClient(cfgAuth);
+
+                fail("Should be rejected.");
+            }
+            catch (ClientAuthenticationException ignored) {}
+
+            checkRejectMetrics(mreg, 1, 1, 2);
+        }
+    }
+
+    /**
+     * Check that failed connection attempts to the grid affect metrics.
+     */
+    @Test
+    public void testClientListenerMetricsRejectGeneral() throws Exception {
+        IgniteConfiguration nodeCfg = getConfiguration()
+            .setClientConnectorConfiguration(new ClientConnectorConfiguration()
+            .setThinClientEnabled(false));
+
+        try (IgniteEx ignite = startGrid(nodeCfg))
+        {
+            MetricRegistry mreg = ignite.context().metric().registry(CLIENT_METRICS);
+
+            checkRejectMetrics(mreg, 0, 0, 0);
+
+            try {
+                Ignition.startClient(getClientConfiguration());
+            }
+            catch (RuntimeException err) {
+                assert err.getMessage().contains("Thin client connection is not allowed");
+            }
+
+            checkRejectMetrics(mreg, 0, 0, 1);
+        }
+    }
+
+    /** */
+    private static ClientConfiguration getClientConfiguration() {
+        return new ClientConfiguration()
+                .setAddresses(Config.SERVER)
+                .setSendBufferSize(0)
+                .setReceiveBufferSize(0);
+    }
+
+    /**
+     * Wait for specific metric to change
+     * @param mreg Metric registry.
+     * @param metric Metric to check.
+     * @param value Metric value to wait for.
+     * @param timeout Timeout.
+     */
+    private void waitForMetricValue(MetricRegistry mreg, String metric, long value, long timeout) {
+        try {
+            GridTestUtils.waitForCondition(new GridAbsPredicate() {
+                @Override public boolean apply() {
+                    return mreg.<IntMetric>findMetric(metric).value() == value;
+                }
+            }, timeout);
+            assertEquals(mreg.<IntMetric>findMetric(metric).value(), value);
+        } catch (IgniteInterruptedCheckedException e) {
+            fail("Interrupted while waiting for metric change");
```

Comment:
Failing this way swallows original exception details. I don't think we need to catch anything at all - let it fail.

Reply:
Done